src/faction.cpp

```diff
@@ -401,7 +401,27 @@ int npc::faction_display( const catacurses::window &fac_w, const int width ) con
         see_color = c_light_red;
     } else if( rl_dist( g->u.pos(), pos() ) > SEEX * 2 || !g->u.sees( pos() ) ) {
         if( u_has_radio && guy_has_radio ) {
-            if( ( g->u.pos().z >= 0 && pos().z >= 0 ) || ( g->u.pos().z == pos().z ) ) {
+            // @TODO : better range calculation than just elevation.
+            int max_range = 200;
+            max_range *= ( 1 + ( g->u.pos().z * 0.1 ) );
+            max_range *= ( 1 + ( pos().z * 0.1 ) );
+            if( is_stationed ) {
+                // if camp that NPC is at, has a radio tower
+                if( stationed_at->has_level( "camp", 20, "[B]" ) ) {
+                    max_range *= 5;
+                }
+            }
+            // if camp that player is at, has a radio tower
+            cata::optional<basecamp *> player_camp = overmap_buffer.find_camp( g->u.global_omt_location().x,
+                    g->u.global_omt_location().y );
+            if( player_camp ) {
+                basecamp *temp_camp = *player_camp;
+                if( temp_camp->has_level( "camp", 20, "[B]" ) ) {
```

Comment:
A cleaner way to do this would be
```C++
if( const cata::optional<basecamp *> player_camp = overmap_buffer.find_camp( g->u.global_omt_location().x, g->u.global_omt_location().y ) ) {
    if( ( *player_camp )->has_level( "camp", 20, "[B]" ) ) {
```